client/src/controllers/LoaderController.ts

```diff
@@ -0,0 +1,130 @@
+ *   data-action="w-loader#activate"
+ * >
+ *  {% icon name="spinner" %}
+ *  <em data-w-loader-target="label">{% trans 'Sign in' %}</em>
+ * </button>
+ */
+export class LoaderController extends Controller {
+  static classes = ['active'];
+  static targets = ['label'];
+  static values = {
+    loading: { type: Boolean, default: false },
+    active: { type: String, default: gettext('Loading') },
+    durationSeconds: { type: Number, default: 30 },
+    label: { type: String, default: '' },
+  };
+
+  declare activeClass: string;
+  declare activeValue: string;
+  declare loadingValue: boolean;
+  declare durationSecondsValue: number;
+  declare labelTarget: HTMLElement;
+  declare hasActiveClass: boolean;
+  declare hasLabelTarget: boolean;
+  declare labelValue: string;
+  timer?: number;
+
+  static afterLoad(identifier: string, application: Application) {
+    const { controllerAttribute } = application.schema;
+    const { actionAttribute } = application.schema;
+
+    const updateLoader = () => {
+      const em = document.querySelector('em') as HTMLElement;
+      document
+        .querySelectorAll(
+          `.button-longrunning:not([data-controller~='${identifier}'])`,
+        )
+        .forEach((button) => {
+          button.setAttribute(controllerAttribute, identifier);
+          button.setAttribute(actionAttribute, 'w-loader#activate');
+
+          if (button.getAttribute('data-clicked-text')) {
+            button.setAttribute(
+              'data-w-loader-active-value',
+              `${button.getAttribute('data-clicked-text')}`,
+            );
+            button.removeAttribute('data-clicked-text');
+          }
+
+          if (em.matches('button em')) {
+            em.setAttribute('data-w-loader-target', 'label');
+          }
+
+          button.setAttribute('data-w-loader-duration-seconds-value', '10');
+        });
+    };
+
+    document.addEventListener('DOMContentLoaded', updateLoader);
+  }
+
+  activate() {
+    // If client-side validation is active on this form, and is going to block submission of the
+    // form, don't activate the spinner
+    const form = this.element.closest('form');
+
+    if (
+      form &&
+      form.checkValidity &&
+      !form.noValidate &&
+      !form.checkValidity()
+    ) {
+      return;
+    }
+
+    window.setTimeout(() => {
+      this.loadingValue = true;
+
+      const durationMs = this.durationSecondsValue * 1000;
+
+      // eslint-disable-next-line no-return-assign
+      this.timer = window.setTimeout(
+        () => (this.loadingValue = false),
+        durationMs,
+      );
+    });
+  }
+
+  loadingValueChanged(isLoading: boolean) {
+    const activeClass = this.hasActiveClass
+      ? this.activeClass
+      : 'button-longrunning-active';
+    this.element.classList.toggle(activeClass, isLoading);
+
+    this.labelValue = this.hasLabelTarget
+      ? (this.labelTarget.textContent as string)
+      : (this.element.textContent as string);
```

Comment:
```suggestion
    if (!this.labelValue) {
      this.labelValue = this.hasLabelTarget
        ? (this.labelTarget.textContent as string)
        : (this.element.textContent as string);
    }
```

A refined approach here - only set the labelValue (the backup of the original text label) if not already set. This way it does not update to the '...loading' (active) text label when loading finishes.